allen-cell-animated/simularium-viewer
# Feature/mesh instancing

Problem
=======
scenes with large numbers of spheres are too slow to render.
Also, the webGL 1 code path and the webgl2 code paths are too entangled in VisGeometry.  The VisGeometry is becoming more and more unreadable.

Solution
========
Optimize mesh rendering by using gpu instancing.
Simplify VisGeometry by refactoring big pieces of rendering code into a LegacyRenderer module.   Also simplify the VisAgent class by having it store an AgentData object of its own.

## Type of change
Please delete options that are not relevant.


Change summary:
---------------
Hopefully this does not break anything but it is a lot of new code.
Cpu javascript processing is becoming more and more the bottleneck for per-frame updates. 
Pdb is the last geometry type that is not using gpu instancing in the renderer yet.

Steps to Verify:
----------------
Just verify that highlighting, following, show/hide, and changing trajectory still works.
Scenes with high counts of meshes should render significantly faster.
A good test scene is `smoldyn_min1_output` or the point test simulator.

Keyfiles (delete if not relevant):
-----------------------
VisAgent, VisGeometry (in particular `updateScene()`), InstancedMesh
LegacyRenderer is entirely new but has code that hopefully will not be run very much and eventually deprecated
Thanks for contributing!
